### ISSUE TITLE
[Feature] 我有一段代码，不知道怎么使用LMDeploy去加速它

### ISSUE THREAD

**Issue by CallmeZhangChenchen** [InternLM/lmdeploy#2958]
### Motivation

工程里面有一段代码，想用LMDeploy加速它
https://github.com/FunAudioLLM/CosyVoice/blob/main/cosyvoice/llm/llm.py

### Related resources

我把它的主体结构截出来了

网络结构，标准的QWen2-0.5B
```
from transformers import Qwen2ForCausalLM
class Qwen2Encoder(torch.nn.Module):
    def __init__(self, pretrain_path):
        super().__init__()
        self.model = Qwen2ForCausalLM.from_pretrained(pretrain_path)

    def forward_one_step(self, xs, masks, cache=None):
        input_masks = masks[:, -1, :]
        outs = self.model(
            inputs_embeds=xs,
            attention_mask=input_masks,
            output_hidden_states=True,
            return_dict=True,
            use_cache=True,
            past_key_values=cache,
        )
        xs = outs.hidden_states[-1]
        new_cache = outs.past_key_values
        
        return xs, new_cache
```

decode 循环
```
       ...
        # 4. cal min/max_length
        min_len = int((text_len - prompt_text_len) * min_token_text_ratio)
        max_len = int((text_len - prompt_text_len) * max_token_text_ratio)

        # 5. step by step decode
        out_tokens = []
        cache = None
        for i in range(max_len):
            begin = time.time()
            y_pred, cache = self.llm.forward_one_step(lm_input,
                                                      masks=torch.tril(torch.ones((1, lm_input.shape[1], lm_input.shape[1]), device=lm_input.device)).to(torch.bool),
                                                      cache=cache)
            print('self.llm.forward_one_step:', time.time() - begin)

            logp = self.llm_decoder(y_pred[:, -1]).log_softmax(dim=-1)
            top_ids = self.sampling_ids(logp.squeeze(dim=0), out_tokens, sampling, ignore_eos=True if i < min_len else False).item()
            if top_ids == self.speech_token_size:
                break
            if top_ids > self.speech_token_size:
                continue
            # in stream mode, yield token one by one
            yield top_ids
            out_tokens.append(top_ids)
            lm_input = self.speech_embedding.weight[top_ids].reshape(1, 1, -1)
```

我想问一下，这段代码可以使用LNDeploy 直接进行加速吗，或者我需要改什么

PS： 我在考虑直接使用LMDeploy 加速一个 forward_one_step （就是一个decode），其他的代码还复用之前的工程结构，
我找到了一个Demo，https://github.com/InternLM/lmdeploy/blob/main/lmdeploy/turbomind/decode.py ，但是decode的参数我不知道怎么传  ^_^～～

**Comment by CallmeZhangChenchen** [comment]
发现了一个问题 Qwen2-0.5 不能直接加载
<img width="1506" alt="image" src="https://github.com/user-attachments/assets/f266193b-72f7-4980-ad01-087863f28d56" />

**Comment by CallmeZhangChenchen** [comment]
这段代码跟传统的语言大模型有几点区别，
1. 直接输入的是 embedding，而不是token，
2. if top_ids > self.speech_token_size: 假如这次decode的结果不符合要求，就复用上次的输入
3. 每次输入都是通过输出id，取好的embedding

**Comment by irexyc** [comment]
不确定能不能做。

decode 返回的是logits，而不是hidden_states。你代码里面的 llm_decoder 似乎不是 model 里面的 lm_head？如果可以 llm_decoder 可以替换 model 里面的 lm_head 的话，pipeliine.get_logits 就相当于 `self.llm.forward_one_step` 和 `self.llm_decoder(y_pred[:, -1])`


关于decode 传 embedding 格式大概是下面这个样子，含义就是有一串 dummy ids，用0表示，然后根据 input_embedding_ranges 用 input_embeddings 来替代 input_ids 经过 lookup table 得到的特征，其中 input_embedding_ranges 里面的 range是左闭右开的。

input_ids: [0,0,0,...] # list 长度是n
input_embeddings # torch.tensor, shape n x hidden_dim
input_embedding_ranges # [[0, n]]

注意到你的 forward_one_step 其实是有状态的 (cache)。sequence_start / sequence_end 也需要设置一下，对于每个 sample 第一次需要设置 sequence_start=True，sequence_end=False，中间的step都设置为False，最后一个step设置为False,True

**Comment by CallmeZhangChenchen** [comment]
感谢 感谢  我将做一个尝试

**Comment by CallmeZhangChenchen** [comment]
@irexyc Hi！多谢指点 ，decode 代码通了
```
    tm_model = tm.TurboMind.from_pretrained(model_path)
    generator = tm_model.create_instance()
    import numpy as np
    input_embeddings = [np.squeeze(np.load('input0.npy'))] # (1, 107, 896)
    input_ids = [0] * 107
    input_embedding_ranges = [[0, 107]]
    logits = generator.decode(input_ids, steps=[0], input_embeddings=input_embeddings,input_embedding_ranges=input_embedding_ranges, sequence_start=True, sequence_end=False)

``` 
但是发现了个问题

```
class Qwen2ForCausalLM：
       def forward():

1165            outputs = self.model(
1166                input_ids=input_ids,
1167                attention_mask=attention_mask,
1168                position_ids=position_ids,
1169                past_key_values=past_key_values,
1170                inputs_embeds=inputs_embeds,
1171                use_cache=use_cache,
1172                output_attentions=output_attentions,
1173                output_hidden_states=output_hidden_states,
1174                return_dict=return_dict,
1175                cache_position=cache_position,
1176            )
1177 
1178            hidden_states = outputs[0]
1181            logits = self.lm_head(hidden_states[:, -num_logits_to_keep:, :])
1182 

1191            return CausalLMOutputWithPast(
1192                loss=loss,
1193                logits=logits,
1194                past_key_values=outputs.past_key_values,
1195                hidden_states=outputs.hidden_states,
1196                attentions=outputs.attentions,
1197            )
```
在  transformers/models/qwen2/modeling_qwen2.py 中 计算 logits = self.lm_head()，入参是 outputs[0]，所以我认为 LMDeploy 也是这样做的


但是在cosvoice 的代码中，计算logp  =  self.llm_decoder()，入参是 outputs.hidden_states[-1][:,-1], 这显然跟 LMDeploy 的计算逻辑不一样
所以我想问一下，LMDeploy 有没有返回特定outputs.hidden_states的接口， 实在没有的话，就只能去改源码了



另外我感觉generator.decode()这个函数，入参都是在 cpu 上的，等最终跑起来之后，应该会有大量的传输开销， 从优化层面有什么方法可以避免嘛

**Comment by irexyc** [comment]
lmdeploy decode 接口返回的是 LLM 模型最终的输出 (output of lm_head)，所以我上面建议你把 llm_decoder 的权重放到 model 的 lm_head里面，这样你用 lmdeploy decode 就相当于做了你上面的 `self.llm.forward_one_step 和 self.llm_decoder(y_pred[:, -1])`

如果你的 lm_input 本身是gpu tensor，想避免传出开销的话，可以试试 pytorch backend，用法类似。

**Comment by CallmeZhangChenchen** [comment]
Hi！ 按照描述，我把llm_decoder 的权重放到 model 的 lm_head里面， 但是他们输出 shape 不一样，结果也不一样
下面是最小化的代码
```
import os
import os.path as osp

import torch
from torch import nn

from lmdeploy import turbomind as tm
from lmdeploy.tokenizer import Tokenizer

from transformers import AutoTokenizer
from transformers import Qwen2ForCausalLM
import numpy as np

os.environ['TM_LOG_LEVEL'] = 'ERROR'

class Qwen2Encoder(torch.nn.Module):
    def __init__(self, pretrain_path):
        super().__init__()
        print('class Qwen2Encoder(torch.nn.Module): pretrain_path', pretrain_path)
        self.model = Qwen2ForCausalLM.from_pretrained(pretrain_path)

    def forward_one_step(self, xs, masks, cache=None):
        input_masks = masks[:, -1, :]
        outs = self.model(
            inputs_embeds=xs,
            attention_mask=input_masks,
            output_hidden_states=True,
            return_dict=True,
            use_cache=True,
            past_key_values=cache,
        )
        xs = outs.hidden_states[-1]
        new_cache = outs.past_key_values

        return xs, new_cache

def hf(model_path):
    qwen2encoder = Qwen2Encoder(model_path)
    lm_input = torch.from_numpy(np.load('input.npy'))
    cache=None
    input_embeddings = torch.from_numpy(np.load('llm_decode.npy'))
    llm_decoder = nn.Linear(896, 6564, bias=False)
    llm_decoder.weight = nn.Parameter(input_embeddings)

    y_pred, cache = qwen2encoder.forward_one_step(lm_input,
            masks=torch.tril(torch.ones((1, lm_input.shape[1], lm_input.shape[1]), device=lm_input.device)).to(torch.bool),
            cache=cache)

    logp = llm_decoder(y_pred[:, -1])             #.log_softmax(dim=-1)
    print(logp.shape, logp)


def main(model_path):
    tm_model = tm.TurboMind.from_pretrained(model_path)
    generator = tm_model.create_instance()

    input_ids = [[]]
    input_embeddings = [np.squeeze(np.load('input.npy'))] # (1, 153, 896)
    input_ids = [0] * 153
    input_embedding_ranges = [[0, 153]]
    logits = generator.decode(input_ids, steps=[0], input_embeddings=input_embeddings,input_embedding_ranges=input_embedding_ranges, sequence_start=True, sequence_end=False)
    print(logits.shape, logits)
    #top_1 = torch.argmax(logits, -1)
    #print(top_1)


if __name__ == '__main__':
    hf('/models/cosyvoicev2/pretrained_models/CosyVoice2-0.5B/CosyVoice-BlankEN')

    main('/models/cosyvoicev2/pretrained_models/CosyVoice2-0.5B/CosyVoice-BlankEN')
```

Qwen2ForCausalLM
```
torch.Size([1, 6564]) tensor([[ 2.1681,  5.4910, -3.0569,  ..., 11.8850, 10.1913,  0.1277]],
       grad_fn=<MmBackward0>)
```

LMDeploy
```
torch.Size([1, 153, 6564]) tensor([[[  7.9552,   2.6536, -11.6485,  ...,  11.9254,  38.8527,  33.6963],
         [ 12.1475,  -0.2788,  -8.6384,  ...,  10.2517,  33.0266,  41.6620],
         [  0.5959,   0.1619,  -2.6343,  ...,   9.7239,   6.1653,  25.0526],
         ...,
         [  2.8050,  -4.3188,   0.0758,  ...,   2.8559,  -0.6115,  31.0399],
         [  4.3672,  -5.3132,  -0.2651,  ...,   2.7728,   1.8311,  25.3163],
         [  2.2029,   5.3990,  -3.1131,  ...,  12.0209,  10.0596,   0.1960]]],
       device='cuda:0')
```









所以我认为这哪里应该有差别
在transformers库中， transformers/models/qwen2/modeling_qwen2.py，Qwen2ForCausalLM的 forward 代码中

```
class Qwen2ForCausalLM(Qwen2PreTrainedModel):
    
     .......

    @add_start_docstrings_to_model_forward(QWEN2_INPUTS_DOCSTRING)
    @replace_return_docstrings(output_type=CausalLMOutputWithPast, config_class=_CONFIG_FOR_DOC)
    def forward(
        self,
        input_ids: torch.LongTensor = None,
        attention_mask: Optional[torch.Tensor] = None,
        position_ids: Optional[torch.LongTensor] = None,
        past_key_values: Optional[List[torch.FloatTensor]] = None,
        inputs_embeds: Optional[torch.FloatTensor] = None,
        labels: Optional[torch.LongTensor] = None,
        use_cache: Optional[bool] = None,
        output_attentions: Optional[bool] = None,
        output_hidden_states: Optional[bool] = None,
        return_dict: Optional[bool] = None,
    ) -> Union[Tuple, CausalLMOutputWithPast]:
        output_attentions = output_attentions if output_attentions is not None else self.config.output_attentions
        output_hidden_states = (
            output_hidden_states if output_hidden_states is not None else self.config.output_hidden_states
        )
        return_dict = return_dict if return_dict is not None else self.config.use_return_dict

        # decoder outputs consists of (dec_features, layer_state, dec_hidden, dec_attn)
        outputs = self.model(
            input_ids=input_ids,
            attention_mask=attention_mask,
            position_ids=position_ids,
            past_key_values=past_key_values,
            inputs_embeds=inputs_embeds,
            use_cache=use_cache,
            output_attentions=output_attentions,
            output_hidden_states=output_hidden_states,
            return_dict=return_dict,
        )

        hidden_states = outputs[0]
        logits = self.lm_head(hidden_states)
        logits = logits.float()

        .......

        return CausalLMOutputWithPast(
            loss=loss,
            logits=logits,
            past_key_values=outputs.past_key_values,
            hidden_states=outputs.hidden_states,
            attentions=outputs.attentions,
        )
```
Qwen2ForCausalLM的 forward 输出有 outputs[0], outputs[1], outputs[2],  outputs.past_key_values, outputs.hidden_states, 
 outputs.attentions,

在需要加速的 python 代码中使用的是  
```
y_pred = outputs.hidden_states[-1]
logp = llm_decoder(y_pred[:, -1]) 
```
我感觉这一块可能有区别

**Comment by irexyc** [comment]
llm_decoder 这个权重的shape是多少，一般 lm_head 是个 linear，shape比如说是 torch.Size([151936, 4096]) # vocab_size x hidden。如果你的 llm_decoder 的 shape 跟原始模型里面的相同的话，直接把这个权重替换 .bin 或者 .safetensors 里面的 lm_head.weight 不就好了。

**Comment by CallmeZhangChenchen** [comment]
llm_decoder  shape 是 torch.Size([6564, 896])， 
原始模型没有lm_head.weight ，它与embed_tokens 共享权重， shape 是 torch.Size([151936, 896])
现在已经得到结果了，您看看这个误差 ok 不

Qwen2ForCausalLM
```
torch.Size([1, 6564]) tensor([[ 2.1681,  5.4910, -3.0569,  ..., 11.8850, 10.1913,  0.1277]],
       grad_fn=<MmBackward0>)
```

LMDeploy 输出最后一维
```
torch.Size([1, 153, 6564]) tensor([[[  7.9552,   2.6536, -11.6485,  ...,  11.9254,  38.8527,  33.6963],
         [ 12.1475,  -0.2788,  -8.6384,  ...,  10.2517,  33.0266,  41.6620],
         [  0.5959,   0.1619,  -2.6343,  ...,   9.7239,   6.1653,  25.0526],
         ...,
         [  2.8050,  -4.3188,   0.0758,  ...,   2.8559,  -0.6115,  31.0399],
         [  4.3672,  -5.3132,  -0.2651,  ...,   2.7728,   1.8311,  25.3163],
         [  2.2029,   5.3990,  -3.1131,  ...,  12.0209,  10.0596,   0.1960]]],
       device='cuda:0')
```

**Comment by irexyc** [comment]
你是改的pytorch backend 里面 qwen2 的代码么？我记得默认 lm_head 的 shape 是和 embedding 的 shape 一样的。

关于误差的话，你可以比较一下 diff，看下相对误差和绝对误差。

**Comment by CallmeZhangChenchen** [comment]
@irexyc  Hi! 打扰了 ，能看下这个问题嘛，我在调用 decode 的时候，单测是没问题的（可以直接返回结果），
但是放到工程里面就出现了这种情况，3 条 INFO，然后过一段时间就崩溃了
<img width="1510" alt="image" src="https://github.com/user-attachments/assets/3dfdcfe0-fd65-44c0-8c6f-3255634939fe" />

**Comment by irexyc** [comment]
@CallmeZhangChenchen

工程里面 decode 会有并发的情况么？

可以把每次的输入存下来看看单独调用有没有问题么，或者找到哪次的输入造成了 segment fault，然后分析下输入。

**Comment by CallmeZhangChenchen** [comment]
@irexyc   感谢    我试出来了，输入input_embeddings 不能放到 gpu 上   ～v～

**Comment by irexyc** [comment]
看到你有一个循环 `for i in range(max_len):` lm_input 和 i 有什么关系么？

**Comment by CallmeZhangChenchen** [comment]
> 看到你有一个循环 `for i in range(max_len):` lm_input 和 i 有什么关系么？
嗯， @irexyc  感谢感谢

第一次循环 lm_input 的维度是 torch.Size([1, 155, 896])
第二次以及之后的每次循环，lm_input是从self.speech_embedding.weight[top_ids].reshape(1, 1, -1)这里取的， 维度是 torch.Size([1, 1, 896]) ，

**Comment by irexyc** [comment]
抱歉没看到下面代码。

不知道你用的是 turbomind backend 还是 pytorch backend。如果是 pytorch backend的话，steps 的设置有问题，这个变量的含义是，比如你有一个很长的序列比如长度是10k，然后每次 decode 1k的长度，那么 steps 分别应该是 [0], [1000], [2000], ... 你填 [i] 是不对的。不过如果你用的是 pytorch backend的话，steps这个参数不起作用，填错的话倒是也没关系。

**Comment by CallmeZhangChenchen** [comment]
上面你好像打错了， 我用的是 turbomind backend，
```
from lmdeploy import turbomind as tm
from lmdeploy import TurbomindEngineConfig
self.tm_model = tm.TurboMind.from_pretrained(pretrain_path, engine_config = TurbomindEngineConfig(max_batch_size=4,cache_max_entry_count=0.2))
self.generator = self.tm_model.create_instance()
self.embed_tokens = torch.nn.Embedding(151936, 896).to('cuda:0')
```

**Comment by CallmeZhangChenchen** [comment]
感谢帮助，感谢 LMDeploy 的优秀团队，问题已经解决 ，close 了  ^-^

**Comment by abc8350712** [comment]
> 感谢帮助，感谢 LMDeploy 的优秀团队，问题已经解决 ，close 了 ^-^

您好，请问是否可以post下完整优化代码呢？
